### DISCUSSION TITLE
working-with-the-result-pattern-part-4-is-the-result-pattern-worth-it/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#440] [Announcements]
# working-with-the-result-pattern-part-4-is-the-result-pattern-worth-it/

In this post I show how to extend your LINQ-based result pattern code to handle more scenarios, such as Task<Result<T>> and collections of Result<T>

https://andrewlock.net/working-with-the-result-pattern-part-4-is-the-result-pattern-worth-it/

**Comment by Gi1gamesh** [comment]
Thanks Andrew, this was a really good series.

**Comment by pjole** [comment]
Hello, I became curious about this pattern but am too lazy to start from the scratch? 

Is there a repo availiable to give me jumpstart?

**Reply by AOne-T** [reply]
As a starter lib you can use https://github.com/vkhorikov/CSharpFunctionalExtensions

**Reply by alexandersorokin** [reply]
I would recommend https://github.com/skbkontur/results . It's lightweight but contains some other features that CSharpFunctionalExtensions doesn't have.

**Comment by nzcoward** [comment]
Some great points tbf, but the methods that Milan is using are pretty basic in the example, and yeah, still not sold on the use of result through e.g. command pipelines.

Having done a fair bit of F# and ocaml, I do quite like the syntax they use, and not a fan of the overhead of a switch extension as in your 2nd post on this topic. 

I am ultimately still working through my own library + syntax, and looking at how I might handle function currying, but I think the syntax might just be too overwhelming, and not overly efficient.

But there is some cool stuff you can do with deconstructing and switch expressions (although pattern matching not supporting custom implicit casting is a bit unfortunate!)

Quick test class showing some different syntax I am currently playing with:
https://gist.github.com/nzcoward/14c1fb3e7230cdee00b38dbfe327a706

**Reply by nzcoward** [reply]
No source ready for the Result or Option classes so it won't build :)

**Comment by AOne-T** [comment]
I want to thank you for this series. It was interesting to learn about LINQ syntax support which looks like promising. 

Also I wanted to share my experience. My acquintance with Result pattern started with Viktor Khorirov's articles (about funcional approach and Railway-Oriented programming) and his library (https://github.com/vkhorikov/CSharpFunctionalExtensions). But our team developed our own internal lib. For now it is highly diverged from what Viktor has, especially in method names (CSharpFuntionalExtensions use more functional naming). 

Our main concern for using this library was to get rid of using exception flow and along with the Maybe pattern (also known as Option) to address the situation with for example `GetById` methods fetching something from DB. At first they returned the entity or null, so we cannot distinguish where there was an error fetching or the entity is not found. Throwing exception caused using try/catch blocks everywhere. So now these methods return `Result<Maybe<TEntity>>` so that we can easily understand what we have.

We have written lots of extension method (for async cases also) which fit our needs. So our code mostly look like this: 
```csharp
        public override async Task<UpdateAttachmentResponse> UpdateAttachment(UpdateAttachmentRequest request, ServerCallContext context)
            => await _serviceProvider.GetRequiredService<UpdateAttachmentRequestValidator>()
                .ValidateToResult(request, _logger)
                .IfValidAsync(() => _attachmentBusiness.GetByIdAsync(request.Id))
                .UnwrapOrFailWithAsync(ErrorDescriptions.ResourceNotFound)
                .EnsureAsync(a => !request.Metadata.AttachmentType.HasValue
                        || EnumMapper.Instance.GetAttachmentMapper().IsDefined(a.EntityInfo.EntityType, request.Metadata.AttachmentType.Value),
                    ValidationErrors.UnsupportedAttachmentType)
                .MapAsync(a => _mapper.Map(request.Metadata, a))
                .IfSuccessAsync(_attachmentBusiness.UpdateAsync)
                .IfAnyAsync(r => new UpdateAttachmentResponse
                {
                    Result = r.ToGrpcResult(_localizer),
                    Item = r.IsSuccess ? _mapper.Map<AttachmentInfo, AttachmentDto>(r.Value) : null
                });
```
Of course sometimes we use the more imperative approach:
```csharp
        ...
        var saveResult = await unitOfWork.SaveChangesAsync();
        if (saveResult.IsFailure)
            return Result.Fail<Entity>(saveResult.Errors);

        var completeResult = unitOfWork.Complete();
        if (completeResult.IsFailure)
            return Result.Fail<Entity>(completeResult.Errors);

        return Result.Ok(entity);
```
We've been working with it for 5 years already and are pretty sure with our initial decision to use Result pattern. With these years devs many devs changed but with some intro any new developer can use this approach within a week

**Reply by nzcoward** [reply]
Big fan of railway-oriented designs, but the pattern is so succinct in actual functional languages and less so in C#.

I don't like the query syntax of linq, but in Andrew's example, it does make it so much cleaner. 

Interesting that you use e.g. UnwrapOrFailWithAsync in your example - I assume that means you have some extension methods that await the task, while others just return it? Or are they still awaited everywhere and wrapped (or maybe you're using ContinueWith?)

**Reply by AOne-T** [reply]
Yes, these are a bunch of extension methods:
```csharp
        public static Task<Result<T>> UnwrapOrFailWithAsync<T>(this Task<Result<Maybe<T>>> resultMaybeTask, Error error) =>
            resultMaybeTask.IfSuccessAsync(maybe => maybe.ToResult(error));

        public static async Task<Result<T>> IfSuccessAsync<T>(this Task<Result> resultTask, Func<T> func, bool continueOnCapturedContext = false) {
            var result = await resultTask.ConfigureAwait(continueOnCapturedContext);
            return result.IfSuccess(func);
        }

        public static Result IfSuccess<T>(in this Result<T> result, Func<T, Result> func) {
            if (result.IsFailure)
                return result.ToResult();
            return func(result.Value);
        }
```
Some just pass through the the result, but others that are at the end of the chain `await` the result.

**Reply by repodn** [reply]
It is probably not a problem to consume functional constructions. But I wouldn't want to write them, and I certainly wouldn't want to face requests from programmers to improve and add to them, preferably immediately. I guess I'm more of a consumer.

**Comment by denisbredikhin** [comment]
Hello Andrew, first of all thank you for the great article.
Can I conclude from this series of articles, that custom exception types in .NET (almost) have no sense? Because it would be an anti-pattern catching them. (ok, maybe only for some specific infrastructure-bases issues they are OK).
But then why we see custom exception types in a LOT of libraries?

**Comment by piller187** [comment]
This is an interesting read. I've started working with Result<> and while it took a min to wrap my head around it I really like how easy it is to flow the methods and read what code is doing. We use method MatchAsync(success, fail) and chain them together 

.MatchAsync(DoNextStep, HandleError)
.MatchAsync(DoAnotherStep, HandleError);

I like the idea of the error handler as you could potentially allow for trying to correct an error and keep going. The trouble I have with this is the params in these methods. The easy way is to make the methods local to the containing method so they can all just use data (like CancellationToken for example) without having to pass in the data. You could also make a "blackboard" type of class in the containing method that each step can use and fill in for next steps but seems a little messier.

The linq style looks cleaner I think but how is error handling done with that?

**Comment by Houuse** [comment]
Jeremy dude doesn't write error handling code because his code doesn't make errors
